src/SecureConnector.php

```diff
@@ -43,13 +43,14 @@ public function connect($uri)
         $context = $this->context;
         $encryption = $this->streamEncryption;
         $connected = false;
+        /** @var array<\React\Promise\PromiseInterface> $promises */
+        $promises = array();
```

Comment:
Diff should be unneeded as setting up a secure connection is strictly sequential and no two promises can exist at the same time. The TLS handshake will start once the TCP/IP connection is ready (no longer pending).

Reply:
This change had 8 hours figuring out what is wrong behind it. And this is probably still not the best solution. But what happens with v3 is that the promise returned from enabling the encryption is never cancelled.

Reply:
Nasty bug that turned out to be a race in the test suite that has been fixed in https://github.com/reactphp/socket/pull/214/files#diff-1ef62b91564dfc6c48899edb5567642889af696e2b7bd1725845b6dfaeb04537L234.